src/Simulation.jl

```diff
@@ -0,0 +1,84 @@
+module Simulation
+
+export Simulation, run!
+
+using Oceananigans.Models
+using Oceananigans.Utils
+
+struct Simulation{M, T, SI, ST, W, R, D, O, P}
+                 model :: M
+                    Δt :: T
+        stop_iteration :: SI
+             stop_time :: ST
+       wall_time_limit :: W
+               runtime :: R
+           diagnostics :: D
+        output_writers :: O
+              progress :: P
+    progress_frequency :: F
+end
+
+function Simulation(model, Δt;
+       stop_iteration = Inf,
+            stop_time = Inf,
+      wall_time_limit = Inf,
+          diagnostics = OrderedDict{Symbol, AbstractDiagnostic}(),
+       output_writers = OrderedDict{Symbol, AbstractOutputWriter}(),
+             progress = nothing,
+   progress_frequency = 1
+   )
+
+   if stop_iteration == Inf && stop_time == Inf && wall_time_limit == Inf
+         @warn "This simulation will run forever as stop iteration = stop time " *
+               "= wall time limit = Inf."
+   end
+
+   runtime = 0.0
+
+   return Simulation(model, Δt, stop_iteration, stop_time, wall_time_limit,
+                     runtime, diagnostics, output_writers, progress)
+end
+
+get_Δt(Δt) = Δt
+get_Δt(wizard::TimeStepWizard) = wizard.Δt
+
+function run!(sim)
+      stop_simulation = false
+
+      if sim.model.clock.iteration > sim.stop_iteration
+            @warn "Simulation will not be run! Model iteration $(sim.model.clock.iteration) " *
+                  "has exceeded simulation stop iteration $(sim.stop_iteration)."
+            stop_simulation = true
+      end
+
+      if sim.model.clock.time > sim.stop_time
+            @warn "Simulation will not be run! Model time $(sim.model.clock.time) " *
+                  "has exceeded simulation stop time $(sim.stop_time)."
+            stop_simulation = true
+      end
+
+      if sim.runtime > sim.wall_time_limit
+            @warn "Simulation will not be run! Simulation run time $(sim.runtime) " *
+                  "has exceeded simulation wall time limit $(sim.wall_time_limit)."
+            stop_simulation = true
+      end
+
+      while !stop_simulation
+            time_before = time()
+
+            time_step!(sim.model, Δt=get_Δt(sim.Δt), Nt=sim.progress_frequency,
```

Comment:
I’m open to changing but my first thought was that we would move the body of time_step here. We can require each TimeStepper to define a function that takes a single time-step. Adams-Bashforth actually needs two such functions, since it has to filter for iteration=0. One reason this is nice is because the “practical” aspects of a simulation — output writing, diagnostics, and other features we add here — are separated from the lower-level time stepping algorithm.